governance/api-reviews.md

```diff
@@ -0,0 +1,207 @@
+imperfections into new APIs for the sake of consistency with old APIs.
+
+### Does it belong in TF at all?
+
+As TF evolves there’s a tendency to put everything inside of it, with costs
+compounding over the long term. If there is a reasonable home for a new API
+outside core TF (say in addons, io, TFP, or other projects entirely) that can be
+strongly preferrable. If new code can be released as independent libraries, it
+should be. This is especially true for APIs that are actively evolving; core TF
+imposes many restrictions, so it’s far better to trial new APIs outside of the
+core library.
+
+## Adding new ops 
+
+Adding new ops to TF should be done with care. We generally prefer not adding
+new ops if possible, but performance, hardware compatibility, and other concerns
+often do require new ops.
+
+When adding new ops, look for:
+ - closure under automatic differentiation (i.e. we avoid ops which are
+   differentiable but not twice-differentiable, or which are technically
+   differentiable but not marked as such)
+ - performant kernels (it’s better not to have an op than to have an op with a
+   suboptimal kernel; we need to make sure kernel experts have reviewed the
+   code)
+ - broadcasting (all numerical ops should broadcast using numpy rules)
+ - does support for this op have to be added to pfor?
+ - dtype support (in general all numerical ops should support the common
+   integer, floating point, and complex dtypes, if they all make sense; we need
+   to watch out for int32 on GPUs though)
+ - device support (cuda kernels should be implemented if possible, and similarly
+   a tf/xla bridge entry should be added if it makes sense)
+ - attributes versus inputs (anything which can be an input to an operation
+   should be an input, and attributes should only be used to parametrize the
+   operation in ways that affect the output dtypes or sometimes shapes)
+ - state management (is the op stateful? Can it instead be made stateless and
+   rely on optimizations like memory reuse for performance? Can it be made to
+   keep its state using one of the existing mechanisms like variables? If not,
+   its state should be encapsulated using resource handles if at all possible)
+ - we generally don’t like ops which are supported only on a single device (be
+   it CPU, GPU, XLA, TPU, etc) and prefer to have at least a plan for writing
+   device-agnostic code
+ - should the python layer for this operation support raggedtensor/sparsetensor?
+
+
+## Keras changes
+
+Keras has its own external API and maintainers, as well as an internal
+team. Keras API changes tend to be cross-cutting as core Keras classes are base
+classes to many TF API endpoints. We generally only look at keras changes after
+a keras owner (say, fchollet or karmel) has already approved it on their
+end. fchollet has guidelines for changes to the Keras API at
+go/keras-api-guidelines. These include many generally good practices that are
```

Comment:
This link should be updated for non-Googlers :)

Reply:
Where this will be documented as Keras will go on a separted org repo https://github.com/tensorflow/community/pull/202?